network/gossip/libp2p/peerManager.go

```diff
@@ -53,14 +54,21 @@ func NewPeerManager(ctx context.Context, logger zerolog.Logger, idsProvider func
 }
 
 // Start kicks off the ambient periodic connection updates
+// It gets blocking till all connection updates are kicking off
 func (pm *PeerManager) Start() error {
-	go pm.updateLoop()
-	go pm.periodicUpdate()
+	wg := &sync.WaitGroup{}
+	wg.Add(2)
+
+	go pm.updateLoop(wg)
+	go pm.periodicUpdate(wg)
+
+	wg.Wait()
 	return nil
 }
```

Comment:
**A flakiness root cause:** Previously, `Start` was not a blocking call. However, tests running it were assuming otherwise. So, there was flakiness due to assuming `updateLoop` and `periodicUpdate` are running right after `Start` returns, which was not always the case.

Reply:
Also, per conversation with @vishalchangrani we plan to make `PeerManager` a `ReadyDoneAware` module: https://github.com/dapperlabs/flow-go/issues/5011

Reply:
I'm a bit uneasy with this as the resolution of the flakiness. Both these goroutines (with an exception in the next paragraph) block immediately after they begin running by entering a `select` statement. As far as I understand the runtime's scheduling behaviour, this goroutine state (blocked on a `select` statement after having been previously scheduled to run on a thread -- let's call it `waiting_after_initial_execution`) is more or less equivalent to the state where the goroutine has been created but not yet scheduled to run on a thread (let's call it `waiting_before_initial_execution`). Using the waitgroup here essentially forces these two goroutines into `waiting_after_initial_execution` rather than `waiting_before_initial_execution`. 

The exception I mentioned earlier is essentially that by forcing these goroutines to be executed, for the `periodicUpdate` method in particular, we _also_ force it to invoke `pm.RequestPeerUpdate()`. I suspect that this might be what is really addressing the flakiness. If this is true I would suggest we simply directly invoke `pm.RequestPeerUpdate()` in the main `Start` method and remove the waitgroup altogether.

Reply:
Fixed in https://github.com/onflow/flow-go/pull/99/commits/c294613989b784fd1104e902ad0eb7b78003dce1